airflow/hooks/hive_hooks.py

```diff
@@ -766,32 +766,47 @@ def get_conn(self, schema=None):
             user=db.login,
             database=schema or db.schema or 'default')
 
-    def get_results(self, hql, schema='default', arraysize=1000):
+    def _get_results(self, hql, schema='default', fetch_size=None):
         from impala.error import ProgrammingError
-        with self.get_conn(schema) as conn:
-            if isinstance(hql, basestring):
-                hql = [hql]
-            results = {
-                'data': [],
-                'header': [],
-            }
-            cur = conn.cursor()
+        if isinstance(hql, basestring):
+            hql = [hql]
+        with self.get_conn(schema) as conn, conn.cursor() as cur:
+            cur.arraysize = fetch_size
             for statement in hql:
                 cur.execute(statement)
-                records = []
-                try:
-                    # impala Lib raises when no results are returned
-                    # we're silencing here as some statements in the list
-                    # may be `SET` or DDL
-                    records = cur.fetchall()
-                except ProgrammingError:
-                    self.log.debug("get_results returned no records")
-                if records:
-                    results = {
-                        'data': records,
-                        'header': cur.description,
-                    }
-            return results
+                # we only get results of statements that returns
+                lowered_statement = statement.lower().strip()
+                if (lowered_statement.startswith('select') or
+                   lowered_statement.startswith('with')):
+                    try:
+                        # impala Lib raises when no results are returned
+                        # we're silencing here as some statements in the list
+                        # may be `SET` or DDL
+                        for row in cur:
+                            yield row
+                    except ProgrammingError:
+                        self.log.debug("get_results returned no records")
+                    description = cur.description
+                    previous_description = self._cursor_description
+                    # if we return data, we expect the schema to be equal
```

Comment:
I'm not sure if users are assuming the same thing at the beginning. Maybe add put the expectation in pydoc?

Reply:
Ya that would be great (I'm one of those who heavily rely on pydoc before using any lib func).